### PR TITLE
PERF: json_normalize, for basic use case

### LINKED ISSUE #15621: PERF: json_normalize
I haven't looked much at the implementation, but guessing simpler cases like this could be optimized.

```python
In [63]: data = [
    ...:     {'name': 'Name',
    ...:      'value': 1.0,
    ...:      'value2': 2.0,
    ...:      'nested': {'a': 'aa', 'b': 'bb'}}] * 1000000

In [64]: %timeit pd.DataFrame(data)
1 loop, best of 3: 847 ms per loop

In [65]: %timeit pd.io.json.json_normalize(data)
1 loop, best of 3: 20 s per loop
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: 1.3.7
pip: 8.1.2
setuptools: 23.0.0
Cython: 0.24.1
numpy: 1.11.2
scipy: 0.18.1
statsmodels: 0.6.1
xarray: 0.8.2
IPython: 5.1.0
sphinx: 1.3.1
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: 1.1.0
tables: 3.2.2
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.2
lxml: 3.6.0
bs4: None
html5lib: 0.999999999
httplib2: 0.9.2
apiclient: 1.5.3
sqlalchemy: 1.0.13
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.40.0
pandas_datareader: 0.2.1
</details>

### PR DESCRIPTION
**Pull request by smpurkis**
- [X] closes #15621
- [X] tests passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Proposed speed up for very simple use cases using the `pd.json_normalize` function. E.g. `pd.json_normalize(data)`

The speed up can be seen in this example:
```
import datetime
import pandas as pd

# example json data
data = {"hello": ["thisisatest", 999898, datetime.date.today()],
        "nest1": {"nest2": {"nest3": "nest3_value", "nest3_int": 3445}},
        "nest1_list": {"nest2": ["blah", 32423, 546456.876, 92030234]},
        "hello2": "string"}

hundred_thousand_rows = [data for i in range(100000)]

s = time()
pd.json_normalize(hundred_thousand_rows)
pandas_json_normalize_time_taken = time() - s
print(f"\npandas time taken for a 100,000 rows: {pandas_json_normalize_time_taken} seconds")
```

With output from Pandas 1.2.2: `pandas time taken for a 100,000 rows: 3.0518009662628174 seconds`
From this branch: `pandas time taken for a 100,000 rows: 0.632451057434082 seconds`

To show tests pass for the appropriate file, ran `pytest pandas/tests/io/json/test_normalize.py -v` [test_normalize.py_pytest.log](https://github.com/pandas-dev/pandas/files/6039361/test_normalize.py_pytest.log)

To show pre-commit passed on file, ran `pre-commit run --files pandas/io/json/_normalize.py` [_normalize.py_pre-commit.log](https://github.com/pandas-dev/pandas/files/6039363/_normalize.py_pre-commit.log)

There was one code check that was caught, running `./ci/code_checks.sh` [code_checks.log](https://github.com/pandas-dev/pandas/files/6039360/code_checks.log)

```
pandas/io/json/_normalize.py:208: error: Incompatible types in assignment (expression has type "List[Union[List[Dict[Any, Any]], Dict[Any, Any]]]", variable has type "Dict[Any, Any]")  [assignment]
```
As it a type hint issue, decided to still make pull request. Can you advice on how to fix, I'm fairly new to type hints?

Kind regards,
Sam

### CODE REVIEW

**Comment by smpurkis** [comment]
Just remembered forgot to update the changelog

File: pandas/io/json/_normalize.py [hunk 1]
```diff
@@ -283,6 +371,14 @@ def _pull_records(js: Dict[str, Any], spec: Union[List, str]) -> List:
     else:
         raise NotImplementedError
 
+    if all(
```
Comment (LuckyDenis) [inline pandas/io/json/_normalize.py:374]: Good afternoon. Why do we need a `for` here? It seems to me that this option is cleaner: `all((record_prefix, meta, meta_prefix, record_prefix, max_level))`

**Review by jreback — CHANGES_REQUESTED** [review]
is it possible to simply dispatch to this if the basic case is selected? 

why is ordering not preserved?

do we have sufficient asv's for this? e.g. pls add the cases you are measuring.

File: pandas/io/json/_normalize.py (hunk 1 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py:374]: That is much simpler than my version, will change it to that at next opportunity

**Comment by smpurkis** [comment]
> is it possible to simply dispatch to this if the basic case is selected?
If possible that would be best, but am not familiar enough with pandas codebase to know where to look. Have tried looking around the `pandas/io/json/` but can't find where dispatch configuration is setup. Any advise for where to look/how to get started would be appreciated.

> why is ordering not preserved?
Oh it is, I need to update that part of the comment.

> do we have sufficient asv's for this? e.g. pls add the cases you are measuring.
I do, I will add those in at the next opportunity and think of a few more cases.


What would you recommend to fix the type hint issue I have?
```
pandas/io/json/_normalize.py:208: error: Incompatible types in assignment (expression has type "List[Union[List[Dict[Any, Any]], Dict[Any, Any]]]", variable has type "Dict[Any, Any]")  [assignment]
```

File: pandas/io/json/_normalize.py (hunk 1 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py:374]: after testing `any` is better here, more specifically 
```
not any((record_prefix, meta, meta_prefix, record_prefix, max_level))
```
as shown by
```
In [1]: t = (None, None)                             

In [2]: all(t)                                       
Out[2]: False

In [3]: t = (None, True)                             

In [4]: all(t)                                       
Out[4]: False

In [5]: t = (None, None)                             

In [6]: any(t)                                       
Out[6]: False

In [7]: t = (None, True)                             

In [8]: any(t)                                       
Out[8]: True
```

**Comment by smpurkis** [comment]
Added asv, as there were none for `json_normalize` function. Although can't run locally as machine isn't powerful enough.
Have also touched up a comment and condition statement

**Comment by smpurkis** [comment]
asv failing, I've not written any before and my laptop is too slow to run them

**Comment by smpurkis** [comment]
asv benchmark should be working now

**Comment by WillAyd** [comment]
Can you run the related JSON benchmarks and post the output of them here?

**Comment by smpurkis** [comment]
Running
```
asv continuous -f 1.1 -E virtualenv upstream/master HEAD -b json
```
gave this log: 
[asv-json-benchmark.log](https://github.com/pandas-dev/pandas/files/6051608/asv-json-benchmark.log)

**Comment by jreback** [comment]
so net effect is this

```
     <issue-15621-improve-json-normalize-perf>       <master>  
         320±2ms        285±0.5ms     0.89  io.json.NormalizeJSON.time_normalize_json('values', 'df_int_floats')
        326±20ms          277±9ms     0.85  io.json.NormalizeJSON.time_normalize_json('records', 'df_int_floats')
```
?

**Comment by jreback** [comment]
this doesnt' seem to match your results.

**Comment by smpurkis** [comment]
> this doesnt' seem to match your results.

I reduced the run from a list of length 100,000 to 10,000 so the benchmark doesn't clog up the avs run. 

Hence you can see the time is ~300 ms, a factor of 10 less time taken

**Comment by smpurkis** [comment]
> this doesnt' seem to match your results.

This benchmark isn't in pandas master only in my forked master, so how are you comparing that?

**Comment by jreback** [comment]
if you added an asv then we could see

this doesn't thatch your timings from the top (the ratio not the absolute time)

**Comment by smpurkis** [comment]
Assuming asv is comparing my forked master to main master then my addition in asv might be wrong. 
Will have another look, and do some reading up on asv

File: pandas/io/json/_normalize.py (hunk 1 above)
Comment (jreback) [inline pandas/io/json/_normalize.py:374]: can you move the comment here and explain what this is checking

File: pandas/io/json/_normalize.py [hunk 2]
```diff
@@ -283,6 +371,14 @@ def _pull_records(js: Dict[str, Any], spec: Union[List, str]) -> List:
...
+        True if x is None else False
```
Comment (jreback) [inline pandas/io/json/_normalize.py:375]: i think you can be explicit, like
```
if (record_path is None and 
    meta is None ....
```

File: pandas/io/json/_normalize.py [hunk 3]
```diff
@@ -121,15 +121,103 @@ def nested_to_record(
...
 def _json_normalize(
-    data: Union[Dict, List[Dict]],
-    record_path: Optional[Union[str, List]] = None,
-    meta: Optional[Union[str, List[Union[str, List[str]]]]] = None,
-    meta_prefix: Optional[str] = None,
-    record_prefix: Optional[str] = None,
-    errors: str = "raise",
-    sep: str = ".",
-    max_level: Optional[int] = None,
+        data: Union[Dict, List[Dict]],
```
Comment (jreback) [inline pandas/io/json/_normalize.py:213]: is there a reason you indented?

File: pandas/io/json/_normalize.py [hunk 4]
```diff
@@ -41,11 +41,11 @@ def convert_to_line_delimits(s):
 
 
 def nested_to_record(
-    ds,
-    prefix: str = "",
-    sep: str = ".",
-    level: int = 0,
-    max_level: Optional[int] = None,
+        ds,
```
Comment (jreback) [inline pandas/io/json/_normalize.py:44]: why the indent?

File: pandas/io/json/_normalize.py [hunk 5]
```diff
@@ -121,15 +121,103 @@ def nested_to_record(
...
+        e.g., for sep='.', { 'foo' : { 'bar' : 0 } } -> foo.bar
+
+    Returns
+    -------
+    frame : DataFrame
+    d - dict or list of dicts, matching `normalised_json_object`
+
+    Examples
+    --------
+    IN[52]: _simple_json_normalize(dict(flat1=1,dict1=dict(c=1,d=2),
```
Comment (jreback) [inline pandas/io/json/_normalize.py:150]: can you use >>> indenting here

File: pandas/io/json/_normalize.py [hunk 6]
```diff
@@ -121,15 +121,103 @@ def nested_to_record(
...
+     'dict1.d': 2,
+     'flat1': 1,
+     'nested.d': 2,
+     'nested.e.c': 1,
+     'nested.e.d': 2}
+
+    """
+
+    # intended as a performance improvement, see #15621
+    def _normalise_json(
```
Comment (jreback) [inline pandas/io/json/_normalize.py:163]: can you make these module level rather than nesting functions

**Review by jreback — CHANGES_REQUESTED** [review]
look pretty good. can you add a whatsnew note in the 1.3 Perf section.

File: pandas/io/json/_normalize.py (hunk 1 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py]: Have updated comment to be more explicit about what it is checking

File: pandas/io/json/_normalize.py (hunk 2 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py]: converted it to this form

File: pandas/io/json/_normalize.py (hunk 4 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py]: reverted

File: pandas/io/json/_normalize.py (hunk 6 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py]: moved functions as advised

File: pandas/io/json/_normalize.py (hunk 5 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py]: converted data structure to a clearer form

File: pandas/io/json/_normalize.py (hunk 3 above)
Comment (smpurkis) [reply pandas/io/json/_normalize.py]: reverted

**Comment by smpurkis** [comment]
Have made the whatsnew note and the changes you advised.
There is still the type hint issue originally picked up, haven't found a way to correct it

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/json/_normalize.py [hunk 7]
```diff
@@ -121,6 +121,98 @@ def nested_to_record(
...
 
 
+def _normalise_json(
+    data: object,
+    key_string: str,
+    new_dict: Dict,
+    separator: str,
+):
+    """
+    Main recursive function
```
[the remark this answers is not in the document]
Comment (smpurkis) [reply pandas/io/json/_normalize.py:131]: Have added the doc-strings to the functions and altered the types, tried to be clearer. Didn't use the re-usable types you suggested in the end, as I think the types I've changed it too are simple enough. Do say if they aren't though

File: pandas/io/json/_normalize.py [hunk 8]
```diff
@@ -121,6 +121,98 @@ def nested_to_record(
...
+        data={k: v for k, v in data.items() if isinstance(v, dict)},
+        key_string="",
+        new_dict={},
+        separator=separator,
+    )
+    return {**top_dict_, **nested_dict_}
+
+
+def _simple_json_normalize(
+    ds: Union[Dict, List[Dict]],
```
[the remark this answers is not in the document]
Comment (smpurkis) [reply pandas/io/json/_normalize.py:167]: same

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @smpurkis very nice!

**Comment by OneMoreSecond** [comment]
@smpurkis 

I meet an exception thrown from
``` python
~/.local/lib/python3.8/site-packages/pandas/io/json/_normalize.py in _normalise_json(data, key_string, normalized_dict, separator)
    150                 # to avoid adding the separator to the start of every key
    151                 key_string=new_key
--> 152                 if new_key[len(separator) - 1] != separator
    153                 else new_key[len(separator) :],
    154                 normalized_dict=normalized_dict,
```

 when running
``` python
import pandas as pd
pd.json_normalize({'': {'a': 1}}, sep='')
```

I cannot understand the issue line. Does it mean?
``` python
if !new_key.startswith(separator)
```

**Pull request merged**